lib/puppet/provider/package/rpm.rb

```diff
@@ -204,200 +188,32 @@ def uninstall_options
-  # as possible.
-  #
-  # returns 1 if str1 is newer than str2,
-  #         0 if they are identical
-  #        -1 if str1 is older than str2
-  def rpmvercmp(str1, str2)
-    return 0 if str1 == str2
-
-    front_strip_re = /^[^A-Za-z0-9~]+/
-
-    while str1.length > 0 or str2.length > 0
-      # trim anything that's in front_strip_re and != '~' off the beginning of each string
-      str1 = str1.gsub(front_strip_re, '')
-      str2 = str2.gsub(front_strip_re, '')
-
-      # "handle the tilde separator, it sorts before everything else"
-      if str1 =~ /^~/ && str2 =~ /^~/
-        # if they both have ~, strip it
-        str1 = str1[1..-1]
-        str2 = str2[1..-1]
-        next
-      elsif str1 =~ /^~/
-        return -1
-      elsif str2 =~ /^~/
-        return 1
-      end
-
-      break if str1.length == 0 or str2.length == 0
-
-      # "grab first completely alpha or completely numeric segment"
-      isnum = false
-      # if the first char of str1 is a digit, grab the chunk of continuous digits from each string
-      if str1 =~ /^[0-9]+/
-        if str1 =~ /^[0-9]+/
-          segment1 = $~.to_s
-          str1 = $~.post_match
-        else
-          segment1 = ''
-        end
-        if str2 =~ /^[0-9]+/
-          segment2 = $~.to_s
-          str2 = $~.post_match
-        else
-          segment2 = ''
-        end
-        isnum = true
-      # else grab the chunk of continuous alphas from each string (which may be '')
-      else
-        if str1 =~ /^[A-Za-z]+/
-          segment1 = $~.to_s
-          str1 = $~.post_match
-        else
-          segment1 = ''
-        end
-        if str2 =~ /^[A-Za-z]+/
-          segment2 = $~.to_s
-          str2 = $~.post_match
-        else
-          segment2 = ''
-        end
-      end
-
-      # if the segments we just grabbed from the strings are different types (i.e. one numeric one alpha),
-      # where alpha also includes ''; "numeric segments are always newer than alpha segments"
-      if segment2.length == 0
-        return 1 if isnum
-        return -1
-      end
-
-      if isnum
-        # "throw away any leading zeros - it's a number, right?"
-        segment1 = segment1.gsub(/^0+/, '')
-        segment2 = segment2.gsub(/^0+/, '')
-        # "whichever number has more digits wins"
-        return 1 if segment1.length > segment2.length
-        return -1 if segment1.length < segment2.length
-      end
-
-      # "strcmp will return which one is greater - even if the two segments are alpha
-      # or if they are numeric. don't return if they are equal because there might
-      # be more segments to compare"
-      rc = segment1 <=> segment2
-      return rc if rc != 0
-    end #end while loop
-
-    # if we haven't returned anything yet, "whichever version still has characters left over wins"
-    if str1.length > str2.length
-      return 1
-    elsif str1.length < str2.length
-      return -1
-    else
-      return 0
-    end
-  end
-
   def insync?(is)
```

Comment:
I think that by modifying `insync?` for rpm we will impact all providers that have rpm as parent. As `insync?` change is in pair with `install` change, to double check if interaction with other providers having rpm as parent is impacted.

For example, a manifest containing a range may work if package in the range is already installed (as `insync?` support ranges), but will not work if the package is not installed (as `install` does not support ranges)

```
grep -r  ':parent => :rpm' *
aptrpm.rb:Puppet::Type.type(:package).provide :aptrpm, :parent => :rpm, :source => :rpm do
rug.rb:Puppet::Type.type(:package).provide :rug, :parent => :rpm do
up2date.rb:Puppet::Type.type(:package).provide :up2date, :parent => :rpm, :source => :rpm do
urpmi.rb:Puppet::Type.type(:package).provide :urpmi, :parent => :rpm, :source => :rpm do
yum.rb:Puppet::Type.type(:package).provide :yum, :parent => :rpm, :source => :rpm do
zypper.rb:Puppet::Type.type(:package).provide :zypper, :parent => :rpm, :source => :rpm do
```